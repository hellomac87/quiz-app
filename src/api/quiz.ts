import { client } from 'src/http/client';
import { QuizParams, QuizResponse } from 'src/types/quiz';

export const getQuiz = (params: QuizParams): Promise<QuizResponse> => client.get('', { params });
