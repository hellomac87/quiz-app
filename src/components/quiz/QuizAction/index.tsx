import Button from 'src/components/common/Button';

import styles from './QuizAction.module.css';

type Props = {
    displayNext: boolean;
    onClickNext: () => void;
    displayLast: boolean;
    onClickLast: () => void;
};

function QuizAction({ displayNext, onClickNext, displayLast, onClickLast }: Props) {
    const handleClick = () => {
        displayNext ? onClickNext() : onClickLast();
    };
    if (!displayNext && !displayLast) {
        return null;
    }
    return (
        <div className={styles.container}>
            <Button onClick={handleClick} variant={displayLast ? 'fill' : 'outline'} fullWidth>
                {displayNext && '다음'}
                {displayLast && '결과 보기'}
            </Button>
        </div>
    );
}

export default QuizAction;
