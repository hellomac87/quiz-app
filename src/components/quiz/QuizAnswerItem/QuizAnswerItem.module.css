.container {
    @apply flex items-center justify-between px-6 py-4 cursor-pointer rounded-lg text-base transition-[background];
    @apply border border-gray-100;
    @apply hover:bg-gray-100;
}
.correct {
    @apply bg-green-200 hover:bg-green-200 text-primary500;
}
.incorrect {
    @apply bg-red-200 hover:bg-red-200 text-red-500;
}
