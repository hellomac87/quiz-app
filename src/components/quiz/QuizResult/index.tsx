import clsx from 'clsx';

import styles from './QuizResult.module.css';

type Props = {
    currentResult?: boolean;
};

function QuizResult({ currentResult }: Props) {
    if (currentResult === undefined) return null;
    return (
        <div
            className={clsx(
                styles.container,
                { [styles.correct]: currentResult },
                { [styles.incorrect]: !currentResult }
            )}
        >
            {currentResult ? 'correct' : 'incorrect'}
        </div>
    );
}

export default QuizResult;
