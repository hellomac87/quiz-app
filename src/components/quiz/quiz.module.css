.question {
    @apply text-4xl text-center mb-12;
}

.answers {
    @apply max-w-lg flex flex-wrap m-auto gap-4;
}

.answer {
    @apply flex items-center justify-center w-[calc(50%-8px)] h-[200px] p-0 rounded-lg bg-white cursor-pointer transition-transform;
    @apply hover:scale-105;

    box-shadow: 20px 20px 60px #d9d9d9, -20px -20px 60px #ffffff;
}
