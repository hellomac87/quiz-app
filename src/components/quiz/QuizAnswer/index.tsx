import { useMemo } from 'react';

import { Quiz } from 'src/types/quiz';
import { suffleQuiz } from 'src/libs/quiz';

import Card from 'src/components/common/Card';

import styles from './QuizAnswer.module.css';

type Props = {
    quiz: Quiz;
    onClickAnswer(answer: string): void;
};

function QuizAnswer({ quiz, onClickAnswer }: Props) {
    const answers = useMemo(() => {
        return suffleQuiz<string>([...quiz.incorrect_answers, quiz.correct_answer]);
    }, [quiz]);

    const handleClickAnswer = (answer: string) => () => {
        onClickAnswer(answer);
    };

    return (
        <div className={styles.container}>
            <h1 className={styles.question}>{quiz.question}</h1>
            <ul className={styles.answers}>
                {answers.map((answer, index) => {
                    const number = `${index + 1}. `;
                    return (
                        <Card key={index} onClick={handleClickAnswer(answer)} className={styles.answer}>
                            {number}
                            {answer}
                        </Card>
                    );
                })}
            </ul>
        </div>
    );
}

export default QuizAnswer;
