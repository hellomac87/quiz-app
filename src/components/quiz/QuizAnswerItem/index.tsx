import { HTMLProps } from 'react';
import clsx from 'clsx';

import { AnswerResult } from 'src/types/quiz';

import styles from './QuizAnswerItem.module.css';

type Props = HTMLProps<HTMLLIElement> & {
    onClick(): void;
    number?: string;
    answer: string;
    correct: AnswerResult;
};

function QuizAnswerItem({ onClick, number, answer, correct, ...rest }: Props) {
    return (
        <li
            onClick={onClick}
            className={clsx(
                styles.container,
                { [styles.correct]: correct === AnswerResult.CORRECT },
                { [styles.incorrect]: correct === AnswerResult.INCORRECT }
            )}
            {...rest}
        >
            <span>
                {number}
                {answer}
            </span>
            <span>
                {correct === AnswerResult.CORRECT && '맞았어요'}
                {correct === AnswerResult.INCORRECT && '틀렸어요'}
            </span>
        </li>
    );
}

export default QuizAnswerItem;
