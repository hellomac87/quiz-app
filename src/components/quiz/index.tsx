import { QuizType } from 'src/types/quiz';

type Props = {
    quiz?: QuizType;
};

function Quiz({ quiz }: Props) {
    if (!quiz) {
        return <div>loading...</div>;
    }

    const answers = suffle<string>([...quiz.incorrect_answers, quiz.correct_answer]);

    const handleClickAnswer = (answer: string) => () => {
        console.log(answer);
    };

    return (
        <div>
            <div>{quiz.category}</div>
            <div>{quiz.question}</div>
            <ul>
                {answers.map((answer, index) => {
                    const number = `${index + 1}. `;
                    return (
                        <li key={index} onClick={handleClickAnswer(answer)}>
                            {number}
                            {answer}
                        </li>
                    );
                })}
            </ul>
        </div>
    );
}

export default Quiz;

function suffle<T>(array: T[]) {
    return array
        .map((value) => ({ value, sort: Math.random() }))
        .sort((a, b) => a.sort - b.sort)
        .map(({ value }) => value);
}
