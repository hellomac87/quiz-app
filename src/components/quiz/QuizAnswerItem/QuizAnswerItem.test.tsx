import renderer from 'react-test-renderer';
import { AnswerResult } from 'src/types/quiz';

import QuizAnswerItem from './index';

const onClick = () => {
    console.log('onClick');
};

const answer = 'Is it correct to write test code like this?';

describe('QuizAnswerItem test', () => {
    test('snapshot: correct', () => {
        const tree = renderer
            .create(<QuizAnswerItem onClick={onClick} answer={answer} correct={AnswerResult.CORRECT} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('snapshot: incorrect', () => {
        const tree = renderer
            .create(<QuizAnswerItem onClick={onClick} answer={answer} correct={AnswerResult.INCORRECT} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('snapshot: none', () => {
        const tree = renderer
            .create(<QuizAnswerItem onClick={onClick} answer={answer} correct={AnswerResult.NONE} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
