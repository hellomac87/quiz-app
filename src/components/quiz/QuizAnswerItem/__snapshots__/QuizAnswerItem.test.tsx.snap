// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizAnswerItem test snapshot: correct 1`] = `
<li
  className="container correct"
  onClick={[Function]}
>
  <span>
    Is it correct to write test code like this?
  </span>
  <span>
    맞았어요
  </span>
</li>
`;

exports[`QuizAnswerItem test snapshot: incorrect 1`] = `
<li
  className="container incorrect"
  onClick={[Function]}
>
  <span>
    Is it correct to write test code like this?
  </span>
  <span>
    틀렸어요
  </span>
</li>
`;

exports[`QuizAnswerItem test snapshot: none 1`] = `
<li
  className="container"
  onClick={[Function]}
>
  <span>
    Is it correct to write test code like this?
  </span>
  <span />
</li>
`;
